process/track/blockProcessor.go

```diff
@@ -0,0 +1,265 @@
+			return selfNotarizedHeadersInfo[i].header.GetNonce() < selfNotarizedHeadersInfo[j].header.GetNonce()
+		})
+	}
+
+	selfNotarizedHeaders := make([]data.HeaderHandler, 0)
+	selfNotarizedHeadersHashes := make([][]byte, 0)
+
+	for _, selfNotarizedHeaderInfo := range selfNotarizedHeadersInfo {
+		selfNotarizedHeaders = append(selfNotarizedHeaders, selfNotarizedHeaderInfo.header)
+		selfNotarizedHeadersHashes = append(selfNotarizedHeadersHashes, selfNotarizedHeaderInfo.hash)
+	}
+
+	return selfNotarizedHeaders, selfNotarizedHeadersHashes
+}
+
+func (bp *blockProcessor) computeLongestChain(shardID uint32, header data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	if check.IfNil(header) {
+		return headers, headersHashes
+	}
+
+	sortedHeaders, sortedHeadersHashes := bp.blockTracker.sortHeadersFromNonce(shardID, header.GetNonce()+1)
+	if len(sortedHeaders) == 0 {
+		return headers, headersHashes
+	}
+
+	longestChainHeadersIndexes := make([]int, 0)
+	headersIndexes := make([]int, 0)
+	bp.getNextHeader(&longestChainHeadersIndexes, headersIndexes, header, sortedHeaders, 0)
+
+	for _, index := range longestChainHeadersIndexes {
+		headers = append(headers, sortedHeaders[index])
+		headersHashes = append(headersHashes, sortedHeadersHashes[index])
+	}
+
+	return headers, headersHashes
+}
+
+func (bp *blockProcessor) getNextHeader(
+	longestChainHeadersIndexes *[]int,
+	headersIndexes []int,
+	prevHeader data.HeaderHandler,
+	sortedHeaders []data.HeaderHandler,
+	index int,
+) {
+	defer func() {
+		if len(headersIndexes) > len(*longestChainHeadersIndexes) {
+			*longestChainHeadersIndexes = headersIndexes
+		}
+	}()
+
+	if check.IfNil(prevHeader) {
+		return
+	}
+
+	for i := index; i < len(sortedHeaders); i++ {
+		currHeader := sortedHeaders[i]
+		if currHeader.GetNonce() > prevHeader.GetNonce()+1 {
+			break
+		}
+
+		if currHeader.GetNonce() == prevHeader.GetNonce()+1 {
+			err := bp.headerValidator.IsHeaderConstructionValid(currHeader, prevHeader)
+			if err != nil {
+				continue
+			}
+
+			err = bp.checkHeaderFinality(currHeader, sortedHeaders, i+1)
+			if err != nil {
+				continue
+			}
+
+			headersIndexes = append(headersIndexes, i)
+			bp.getNextHeader(longestChainHeadersIndexes, headersIndexes, currHeader, sortedHeaders, i+1)
+			headersIndexes = headersIndexes[:len(headersIndexes)-1]
+		}
+	}
+}
+
+func (bp *blockProcessor) checkHeaderFinality(
+	header data.HeaderHandler,
+	sortedHeaders []data.HeaderHandler,
+	index int,
+) error {
+
+	if check.IfNil(header) {
+		return process.ErrNilBlockHeader
+	}
+
+	prevHeader := header
+	numFinalityAttestingHeaders := uint64(0)
+
+	for i := index; i < len(sortedHeaders); i++ {
+		currHeader := sortedHeaders[i]
+		if numFinalityAttestingHeaders >= bp.blockFinality || currHeader.GetNonce() > prevHeader.GetNonce()+1 {
+			break
+		}
+
+		if currHeader.GetNonce() == prevHeader.GetNonce()+1 {
```

Comment:
the same check is inside IsHeaderConstructionValid - so this is not needed.

Reply:
I just wanted to avoid the consumed time with the other checks which are made before the nonce check in headerValidator. Now I checked again the IsHeaderConstructionValid method and it checks only the round before the nonce so I removed this check here.